README.md
<h1 align="center">Robo Kit 🤖</h1>

<div align="center">
  <strong>
    Build custom firmware for simple robotic automation
  </strong>
</div>

## Modules

- [`robokit`](./robokit) : [![Crates.io version](https://img.shields.io/crates/v/robokit.svg?style=flat-square) ](https://crates.io/crates/robokit)  [![Download](https://img.shields.io/crates/d/robokit.svg?style=flat-square)](https://crates.io/crates/robokit)  [![docs.rs docs](https://img.shields.io/badge/docs-latest-blue.svg?style=flat-square)](https://docs.rs/robokit)

## Firmwares

- [`blinky`](./blinky)
- [`gridbot-tahi`](./gridbot-tahi)

## About

The short-term goal is to build [an automated machine for grid beam production](https://github.com/villagekit/gridbot-tahi).

The long-term goal is to provide a ([real-time interrupt-driven](https://rtic.rs)) [actor-based](https://en.wikipedia.org/wiki/Actor_model) foundation for robotic automation or CNC machine control.

If you're here and like what's happening, please give this a star and [say hi](https://github.com/villagekit/robokit/issues)! 👋

## Features

- Minimal
  - Designed for `no-std` environments
- Extensible
  - Setup your robot with your own actuators with your own names
      - E.g. Isn't limited to only x, y, z linear axes
- Command system (like G-Code)
  - Run a sequence of commands (one at a time)
  - Run setup commands at beginning and/or teardown commands at end (in parallel)
- Actuators:
  - [x] Led
      - Actions:
        - Set { is_on }
        - Blink { duration }
  - [x] Linear Axis
      - Drivers: [Stepper](https://github.com/braun-embedded/stepper)
      - Actions:
        - MoveRelative { max_acceleration, distance }
        - MoveAbsolute { max_acceleration, position }
        - Home { max_acceleration, back_off_distance }
  - [x] Spindle
      - Drivers:
        - JmcHsv57
      - Actions:
        - Set(On { rpm })
        - Set(Off)
  - [ ] Relay (Pneumatic actuator)
- Sensors
  - [x] Input switch
      - Button
      - Limit switch
  - [ ] Rotary encoder
  - [ ] Linear encoder
- Interfaces
  - [ ] Physical controls
  - [ ] JSON-RPC
  - [ ] Web

## Example

[`./blinky/src/main.rs`](./blinky/src/main.rs)

(for [Nucleo-F767ZI](https://nz.element14.com/stmicroelectronics/nucleo-f767zi/dev-board-nucleo-32-mcu/dp/2546569))

```rust
#![no_main]
#![no_std]

use blinky as _;

use core::task::Poll;
use cortex_m_rt::entry;
use defmt::Debug2Format;
use fugit::ExtU32;
use robokit::{
    actuator_set, Command, LedAction, LedDevice, RobotBuilder, Sensor, SuperTimer, SwitchDevice,
    SwitchStatus,
};
use stm32f7xx_hal::{pac, prelude::*};

use blinky::init_heap;

const TICK_TIMER_HZ: u32 = 1_000_000;
const ACTIVE_COMMANDS_COUNT: usize = 1;

actuator_set!(
    Led { Green, Blue, Red },
    LedAction<TICK_TIMER_HZ>,
    LedId,
    LedSet,
    LedSetError
);

fn get_run_commands<const TIMER_HZ: u32>() -> [Command<TIMER_HZ, LedId, (), ()>; 6] {
    [
        Command::Led(
            LedId::Green,
            LedAction::Blink {
                duration: 50.millis(),
            },
        ),
        Command::Led(
            LedId::Blue,
            LedAction::Blink {
                duration: 100.millis(),
            },
        ),
        Command::Led(
            LedId::Red,
            LedAction::Blink {
                duration: 200.millis(),
            },
        ),
        Command::Led(
            LedId::Red,
            LedAction::Blink {
                duration: 50.millis(),
            },
        ),
        Command::Led(
            LedId::Blue,
            LedAction::Blink {
                duration: 100.millis(),
            },
        ),
        Command::Led(
            LedId::Green,
            LedAction::Blink {
                duration: 200.millis(),
            },
        ),
    ]
}

#[entry]
fn main() -> ! {
    init_heap();

    defmt::println!("Init!");

    let p = pac::Peripherals::take().unwrap();

    let rcc = p.RCC.constrain();
    let clocks = rcc.cfgr.sysclk(216.MHz()).freeze();

    let gpiob = p.GPIOB.split();
    let gpioc = p.GPIOC.split();

    let tick_timer = p.TIM5.counter_us(&clocks);
    let mut super_timer = SuperTimer::new(tick_timer, u32::MAX);

    let user_button_pin = gpioc.pc13.into_floating_input();
    let user_button_timer = super_timer.sub();
    let mut user_button = SwitchDevice::new_active_high(user_button_pin, user_button_timer);

    let green_led_pin = gpiob.pb0.into_push_pull_output();
    let green_led_timer = super_timer.sub();
    let green_led = LedDevice::new(green_led_pin, green_led_timer);

    let blue_led_pin = gpiob.pb7.into_push_pull_output();
    let blue_led_timer = super_timer.sub();
    let blue_led = LedDevice::new(blue_led_pin, blue_led_timer);

    let red_led_pin = gpiob.pb14.into_push_pull_output();
    let red_led_timer = super_timer.sub();
    let red_led = LedDevice::new(red_led_pin, red_led_timer);

    let mut robot = RobotBuilder::new()
        .with_leds(LedSet::new(green_led, blue_led, red_led))
        .build()
        .with_run_commands(&get_run_commands())
        .build::<ACTIVE_COMMANDS_COUNT>();

    super_timer.setup().expect("Failed to setup super time");
    loop {
        super_timer.tick().expect("Failed to tick super timer");

        if let Some(user_button_update) = user_button.sense().expect("Error reading user button") {
            if let SwitchStatus::On = user_button_update.status {
                robot.toggle();
            }
        }

        if let Poll::Ready(Err(err)) = robot.poll() {
            defmt::println!("Unexpected error: {}", Debug2Format(&err));

            robot.stop();
        }
    }
}
```

## Development

See [`docs/dev.md`](./docs/dev.md)

## License

Copyright 2023 Village Kit Limited

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

with help from [`git log`](https://www.git-scm.com/docs/git-log):

```shell
git log --oneline --format="- [%h](https://github.com/villagekit/robokit/commit/%H): %s"
```

## [robokit-0.3.0](https://github.com/villagekit/robokit/releases/tag/robokit-0.3.0)

- [af19fde](https://github.com/villagekit/robokit/commit/af19fde814b6773133d90d4f5d59cc153491d963): cleanup modules (#42)
- [ad4b3df](https://github.com/villagekit/robokit/commit/ad4b3df0c8d6aa3bdc7e630e83af936eaeb64fe9): back to actuator sets (#41)
- [d7bfb94](https://github.com/villagekit/robokit/commit/d7bfb949bd68cfe5e4e6bfe389e390d4bdc7fa25): update for gridbot tahi demo (#37)
- [6029690](https://github.com/villagekit/robokit/commit/60296909ef4f9f0d3e00096d21d2bcab1fc4d6c7): fix silly runtime bugs (#40)
- [5b9f674](https://github.com/villagekit/robokit/commit/5b9f67459be12d53be93e9d88beb63d793a2f27a): refactor switch errors (#39)
- [5f71470](https://github.com/villagekit/robokit/commit/5f71470eb3655a9c8c14693b8ebf3e73df21e2af): use fixed_map as better actuator registry (#38)
- [eb28d6d](https://github.com/villagekit/robokit/commit/eb28d6d8cb577561ccd440dea8208c58673ab276): validate all command ids exist when building robot (#36)
- [5acbe2a](https://github.com/villagekit/robokit/commit/5acbe2a90b2abcce3eb23d3d9531c3c33df9deaa): refactor axis errors (#35)
- [3affaa4](https://github.com/villagekit/robokit/commit/3affaa4465e4a43f402d523833e7b9bb89f8ac96): refactor led (#34)

## [robokit-0.2.0](https://github.com/villagekit/robokit/releases/tag/robokit-0.2.0)

- [a4832dc](https://github.com/villagekit/robokit/commit/a4832dc78d6c03d637b7fb0c4584d36336d31679): use explicit heapless sizes (#33)
- [55c3263](https://github.com/villagekit/robokit/commit/55c3263137d6ce35e39c21e3301d5d2a75b43e50): super time 2 (#32)

## [robokit-0.1.0](https://github.com/villagekit/robokit/releases/tag/robokit-0.1.0)

- [d4a7c2c](https://github.com/villagekit/robokit/commit/d4a7c2c73e912032651dbdcbd2db395f3e92ccd2): too many keywords, sorry crates.io
- [9c74370](https://github.com/villagekit/robokit/commit/9c743701ffdde27651eaac5dd62d876521d2f806): step by step libraryify 3 (#31)
- [8004ceb](https://github.com/villagekit/robokit/commit/8004cebdc23ac7c7f4f6a9e419358ef2f84556a7): step by step libraryify 2 (#30)
- [67c93e7](https://github.com/villagekit/robokit/commit/67c93e74af2a7ed62e0c205d499a50b555fcbd19): step by step libraryify (#28)
- [afbaa72](https://github.com/villagekit/robokit/commit/afbaa729afb47ae7ca0a103baa34819949ebc9d7): add axis homing (#20)
- [dde54d3](https://github.com/villagekit/robokit/commit/dde54d337aa03030b8ea17d65f2896f866ea3770): generic traits (#27)
- [c645309](https://github.com/villagekit/robokit/commit/c645309cb22a25e1e9f6cb99f2cb99a92d47062d): tidy: remove rtic, upgrade crates, add heap alloc (#26)
- [e23970a](https://github.com/villagekit/robokit/commit/e23970afc631e851ddd3267ae73a265af4013a06): upgrade deps, use fugit time with stepper (#24)
- [bb424c1](https://github.com/villagekit/robokit/commit/bb424c13896f65a5ed9d83a582f8d06d22691d8d): update deps to latest versions
- [0d73356](https://github.com/villagekit/robokit/commit/0d733562c8db8be5ff5cfa3843e98d6a149ada29): use fugit time with stepper
- [037ec1c](https://github.com/villagekit/robokit/commit/037ec1c9af785bd104999f210b63c2b77a799191): super time (#25)
- [3d04058](https://github.com/villagekit/robokit/commit/3d040587fad06fe89250f9fa0ed54c429e2e204f): improve command flow (#23)
- [442c1ca](https://github.com/villagekit/robokit/commit/442c1caa515f4c2608a50cbf4ddb3646d05f8bb6): use merged `stm32f7xx-hal` version
- [e953ea8](https://github.com/villagekit/robokit/commit/e953ea86dfafb2878f1899340c51fe38c45a5d78): use published `rmodbus@0.6.1`
- [6fabab0](https://github.com/villagekit/robokit/commit/6fabab0ce2f1a47327714a78a7c5d2fa8fb40e51): iHSV57 servo spindle: part 2 (#21)
- [0e0dfb5](https://github.com/villagekit/robokit/commit/0e0dfb5e16aa395ef0e8f225346c610b7472b171): iHSV57 servo spindle (#19)
- [6390f37](https://github.com/villagekit/robokit/commit/6390f37e3d649d41216190ec022aae710223fd2a): add switch sensor (#15)
- [6b7a74d](https://github.com/villagekit/robokit/commit/6b7a74d5c294d3cfe08c439307a18c26997f5065): update docs and add license (#13)
- [3ea451e](https://github.com/villagekit/robokit/commit/3ea451ee20c6d049e157b2449d1972b2e56a9ff8): stepper axis (#12)
- [5bfe5df](https://github.com/villagekit/robokit/commit/5bfe5dfb2f5d14013f8a2f9e896ef4997572ae83): remove unused src/error.rs
- [9412bec](https://github.com/villagekit/robokit/commit/9412becce3c0930da490fb8574a9f30c3e4b7a1c): implement command system (#11)
- [e954227](https://github.com/villagekit/robokit/commit/e95422701247898d141c6a26b7e2b9048c795047): Merge pull request #10 from villagekit/rusty-revival
- [0ca89fc](https://github.com/villagekit/robokit/commit/0ca89fcf7376f23aadb1c7b22e896410db289211): follow latest knurling-rs/app-template
- [8ffbd1d](https://github.com/villagekit/robokit/commit/8ffbd1dfb69f47f709a1aba23c3d323e82581079): a fresh new beginning

## C++ prototype

- [1405407](https://github.com/villagekit/robokit/commit/140540713791b93767f0ec5e1bf4c4d4946cc533): improve stepper motor code (#6)
- [e06dc44](https://github.com/villagekit/robokit/commit/e06dc446741e124882bc86e3e64efc9675cbd1f4): interrupt safe queue (#5)
- [179f766](https://github.com/villagekit/robokit/commit/179f7667845aec531c1827cac82088bff07fb72f): remove variant.hpp (in ./3rdparty)
- [7f125fa](https://github.com/villagekit/robokit/commit/7f125facf64083ccba83933e72ba5ca0a77f11a1): add dev instructions to readme
- [1e4a697](https://github.com/villagekit/robokit/commit/1e4a69716e340628aced00d3e59924d03af1d0ea): pulse motors (#4)
- [2a91413](https://github.com/villagekit/robokit/commit/2a9141305d36a83cff89de3b0e66dbb37d7107f0): send json data to browser and render (#3)
- [3dd87a4](https://github.com/villagekit/robokit/commit/3dd87a4a9e1eebfee3566db285892fcb6f4bb36e): clean up code (#2)

src/main.rs
#![no_main]
#![no_std]

use gridbot as _;

use core::task::Poll;
use cortex_m_rt::entry;
use defmt::Debug2Format;
use fugit::ExtU32;
use stm32f7xx_hal::{
    gpio::{self, Alternate, Floating, Input, Output, Pin, PullUp, PushPull},
    pac,
    prelude::*,
    rcc::BusTimerClock,
    serial::{Config as SerialConfig, DataBits as SerialDataBits, Parity as SerialParity, Serial},
    timer::Counter,
    watchdog,
};

use gridbot::{
    actuators::{
        axis::{AxisDevice, AxisLimitSide},
        led::LedDevice,
        spindle::{SpindleDevice, SpindleDriverJmcHsv57},
    },
    init_heap,
    machine::Machine,
    runner::{Runner, RunnerAxes, RunnerLeds, RunnerSpindles},
    sensors::{
        switch::{SwitchActiveHigh, SwitchActiveLow, SwitchDevice, SwitchStatus},
        Sensor,
    },
    timer::{setup as timer_setup, tick as timer_tick, SubTimer, TICK_TIMER_HZ},
};

pub const TICK_TIMER_MAX: u32 = u32::MAX;
pub type TickTimer = Counter<pac::TIM5, TICK_TIMER_HZ>;

type UserButtonPin = Pin<'C', 13, Input<Floating>>;
type UserButtonTimer = SubTimer;
// type UserButtonError = SwitchError<<UserButtonPin as InputPin>::Error>;
type UserButton = SwitchDevice<UserButtonPin, SwitchActiveHigh, UserButtonTimer, TICK_TIMER_HZ>;

/* actuators */

type GreenLedPin = Pin<'B', 0, Output<PushPull>>;
type GreenLedTimer = SubTimer;
type BlueLedPin = Pin<'B', 7, Output<PushPull>>;
type BlueLedTimer = SubTimer;
type RedLedPin = Pin<'B', 14, Output<PushPull>>;
type RedLedTimer = SubTimer;

const X_AXIS_TIMER_HZ: u32 = 1_000_000;
type XAxisDirPin = Pin<'G', 9, Output<PushPull>>; // D0
type XAxisStepPin = Pin<'G', 14, Output<PushPull>>; // D1
type XAxisTimer = Counter<pac::TIM3, X_AXIS_TIMER_HZ>;

type MainSpindleSerial = Serial<pac::USART2, (gpio::PD5<Alternate<7>>, gpio::PD6<Alternate<7>>)>;
type MainSpindleDriver = SpindleDriverJmcHsv57<MainSpindleSerial>;

/* sensors */
type XAxisLimitMinPin = Pin<'F', 15, Input<PullUp>>; // D2
type XAxisLimitMinTimer = SubTimer;
type XAxisLimitMin =
    SwitchDevice<XAxisLimitMinPin, SwitchActiveLow, XAxisLimitMinTimer, TICK_TIMER_HZ>;
type XAxisLimitMaxPin = Pin<'F', 14, Input<PullUp>>; // D4
type XAxisLimitMaxTimer = SubTimer;
type XAxisLimitMax =
    SwitchDevice<XAxisLimitMaxPin, SwitchActiveLow, XAxisLimitMaxTimer, TICK_TIMER_HZ>;

#[entry]
fn main() -> ! {
    init_heap();

    defmt::println!("Init!");

    let p = pac::Peripherals::take().unwrap();

    let rcc = p.RCC.constrain();
    let clocks = rcc.cfgr.sysclk(216.MHz()).freeze();

    let gpiob = p.GPIOB.split();
    let gpioc = p.GPIOC.split();
    let gpiod = p.GPIOD.split();
    let gpiof = p.GPIOF.split();
    let gpiog = p.GPIOG.split();

    let mut tick_timer: TickTimer = p.TIM5.counter_us(&clocks);
    timer_setup(&mut tick_timer, TICK_TIMER_MAX).unwrap();

    let user_button_pin: UserButtonPin = gpioc.pc13.into_floating_input();
    let user_button_timer: UserButtonTimer = SubTimer::new();
    let mut user_button: UserButton = SwitchDevice::new(user_button_pin, user_button_timer);

    let green_led_pin: GreenLedPin = gpiob.pb0.into_push_pull_output();
    let green_led_timer: GreenLedTimer = SubTimer::new();
    let green_led = LedDevice::new(green_led_pin, green_led_timer);
    let blue_led_pin: BlueLedPin = gpiob.pb7.into_push_pull_output();
    let blue_led_timer: BlueLedTimer = SubTimer::new();
    let blue_led = LedDevice::new(blue_led_pin, blue_led_timer);
    let red_led_pin: RedLedPin = gpiob.pb14.into_push_pull_output();
    let red_led_timer: RedLedTimer = SubTimer::new();
    let red_led = LedDevice::new(red_led_pin, red_led_timer);

    defmt::println!(
        "Stepper timer clock: {}",
        <pac::TIM3 as BusTimerClock>::timer_clock(&clocks)
    );

    let max_acceleration_in_millimeters_per_sec_per_sec = 20_f64;

    // https://www.makerstore.com.au/product/gear-m1/
    let steps_per_revolution = 6400_f64;
    let millimeters_per_revolution = 125.66_f64;
    let steps_per_millimeter = steps_per_revolution * (1_f64 / millimeters_per_revolution);

    defmt::println!("Steps per mm: {}", steps_per_millimeter);

    let x_axis_dir_pin: XAxisDirPin = gpiog.pg9.into_push_pull_output();
    let x_axis_step_pin: XAxisStepPin = gpiog.pg14.into_push_pull_output();
    let x_axis_timer: XAxisTimer = p.TIM3.counter(&clocks);
    let x_axis_limit_min_pin: XAxisLimitMinPin = gpiof.pf15.into_pull_up_input();
    let x_axis_limit_min_timer: XAxisLimitMinTimer = SubTimer::new();
    let x_axis_limit_min: XAxisLimitMin =
        SwitchDevice::new_active_low(x_axis_limit_min_pin, x_axis_limit_min_timer);
    let x_axis_limit_max_pin: XAxisLimitMaxPin = gpiof.pf14.into_pull_up_input();
    let x_axis_limit_max_timer: XAxisLimitMaxTimer = SubTimer::new();
    let x_axis_limit_max: XAxisLimitMax =
        SwitchDevice::new_active_low(x_axis_limit_max_pin, x_axis_limit_max_timer);
    let x_axis = AxisDevice::new_dq542ma(
        x_axis_dir_pin,
        x_axis_step_pin,
        x_axis_timer,
        max_acceleration_in_millimeters_per_sec_per_sec,
        steps_per_millimeter,
        x_axis_limit_min,
        x_axis_limit_max,
        AxisLimitSide::Min,
    );

    let main_spindle_serial_tx = gpiod.pd5.into_alternate();
    let main_spindle_serial_rx = gpiod.pd6.into_alternate();
    let main_spindle_serial: MainSpindleSerial = Serial::new(
        p.USART2,
        (main_spindle_serial_tx, main_spindle_serial_rx),
        &clocks,
        SerialConfig {
            baud_rate: 57600.bps(),
            data_bits: SerialDataBits::Bits9,
            parity: SerialParity::ParityEven,
            ..Default::default()
        },
    );
    let main_spindle_driver: MainSpindleDriver = SpindleDriverJmcHsv57::new(main_spindle_serial);
    let main_spindle = SpindleDevice::new(main_spindle_driver);

    let runner = Runner::new(
        RunnerLeds {
            green_led,
            blue_led,
            red_led,
        },
        RunnerAxes { x_axis },
        RunnerSpindles { main_spindle },
    );
    let mut machine = Machine::new(runner);

    let mut iwdg = watchdog::IndependentWatchdog::new(p.IWDG);

    iwdg.start(2.millis());

    loop {
        timer_tick(&mut tick_timer, TICK_TIMER_MAX).unwrap();

        if let Some(user_button_update) = user_button.sense().expect("Error reading user button") {
            if let SwitchStatus::On = user_button_update.status {
                machine.toggle();
            }
        }

        if let Poll::Ready(Err(err)) = machine.poll() {
            defmt::println!("Unexpected error: {}", Debug2Format(&err));

            machine.stop();
            loop {
                match machine.poll() {
                    _ => {}
                }
            }
        }

        iwdg.feed();
    }
}
